internal/overlord/cmdstate/handlers.go

```diff
@@ -0,0 +1,517 @@
+	var stdin *os.File
+	var stdout *os.File
+	var stderr *os.File
+
+	// Closed to make the controlLoop exit early.
+	controlExit := make(chan struct{})
+	defer close(controlExit)
+
+	pidCh := make(chan int)
+	childDead := make(chan struct{})
+	var wgOutputSent sync.WaitGroup
+
+	if e.useTerminal {
+		var uid, gid int
+		if e.userID != nil && e.groupID != nil {
+			uid, gid = *e.userID, *e.groupID
+		} else {
+			uid = os.Getuid()
+			gid = os.Getgid()
+		}
+
+		master, slave, err := ptyutil.OpenPty(int64(uid), int64(gid))
+		if err != nil {
+			return err
+		}
+		afterClosers = append(afterClosers, master)
+		beforeClosers = append(beforeClosers, slave)
+
+		stdin = slave
+		stdout = slave
+		stderr = slave // stderr will be overwritten below if splitStderr true
+
+		if e.width > 0 && e.height > 0 {
+			err = ptyutil.SetSize(int(master.Fd()), e.width, e.height)
+			if err != nil {
+				logger.Noticef("Exec %s: cannot set initial terminal size to %dx%d: %v",
+					task.ID(), e.width, e.height, err)
+			}
+		}
+
+		go e.controlLoop(task.ID(), pidCh, controlExit, int(master.Fd()))
+
+		// Start goroutine to mirror PTY I/O to "stdio" websocket.
+		wgOutputSent.Add(1)
+		go func() {
+			defer wgOutputSent.Done()
+
+			logger.Debugf("Exec %s: started mirroring websocket", task.ID())
+			defer logger.Debugf("Exec %s: finished mirroring websocket", task.ID())
+
+			ioConn := e.getWebsocket(wsStdio)
+			defer func() {
+				_ = ioConn.Close()
+			}()
+
+			readDone, writeDone := wsutil.WebsocketExecMirror(
+				ioConn, master, master, childDead, int(master.Fd()))
+			<-readDone
+			<-writeDone
+		}()
+	} else {
+		go e.controlLoop(task.ID(), pidCh, controlExit, -1)
+
+		// Start goroutine to receive stdin from "stdio" websocket and write to
+		// cmd.Stdin pipe.
+		ioConn := e.getWebsocket(wsStdio)
+		stdinReader, stdinWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		stdin = stdinReader
+		afterClosers = append(afterClosers, stdinReader)
+		beforeClosers = append(beforeClosers, stdinWriter)
+		go func() {
+			<-wsutil.WebsocketRecvStream(stdinWriter, ioConn)
+			stdinWriter.Close()
+		}()
+
+		// Start goroutine to receive from cmd.Stdout pipe and write to "stdio"
+		// websocket.
+		stdoutReader, stdoutWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		afterClosers = append(afterClosers, stdoutReader)
+		beforeClosers = append(beforeClosers, stdoutWriter)
+		stdout = stdoutWriter
+		stderr = stdoutWriter // stderr will be overwritten below if splitStderr true
+		wgOutputSent.Add(1)
+		go func() {
+			defer wgOutputSent.Done()
+			<-wsutil.WebsocketSendStream(ioConn, stdoutReader, -1)
+			stdoutReader.Close()
+		}()
+	}
+
+	if e.splitStderr {
+		// Start goroutine to receive from cmd.Stderr pipe and write to a
+		// separate "stderr" websocket.
+		stderrReader, stderrWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		afterClosers = append(afterClosers, stderrReader)
```

Comment:
In theory I suppose we wouldn't need this here, as it's being closed below and there's wgOutputSent blocking the termination before that takes place. But that shouldn't be a problem.. belt and suspender approach I suppose.